blog/content/post/2020-12-08-IOMMU.md

```diff
@@ -0,0 +1,87 @@
+layout: post
+published: true
+date: 2020-12-08
+archives: "2020"
+
+tags:
+  - Firmware
+  - IOMMU
+  - Security
+categories:
+  - Firmware
+  - Security
+
+---
+
+## Introduction
+
+The I/O memory management unit (IOMMU) is a type of memory management unit (MMU)
+that connects a Direct Memory Access (DMA) capable expansion bus to the main memory.
+It extends the system architecture by adding support for the virtualization of
+memory addresses used by peripheral devices. Additionally, it provides memory
+isolation and protection by enabling system software to control which areas of
+physical memory an I/O device may access. It also helps filter and remap interrupts
+from peripheral devices. 
+
+Advantages of IOMMU usage:
+
+  - One single contiguous virtual memory region can be mapped to multiple non-contiguous
+  physical memory regions. IOMMU can make a non-contiguous memory region appear
+  contiguous to a device (scatter/gather). Scatter/gather optimizes streaming DMA
+  performance for the I/O device.
+  - Memory isolation and protection: device can only access memory regions that are
+  mapped for it. Hence faulty and/or malicious devices can't corrupt memory.
+  - Memory isolation allows safe device assignment to a virtual machine without
+  compromising host and other guest OSes.
+  - IOMMU enables 32-bit DMA capable devices to access to > 4GB memory.
+  - Support hardware interrupt remapping. It extends limited hardware interrupts
+  to software interrupts. Primary uses are the interrupt isolation and translation
+  between interrupt domains (E.g. ioapic vs x2apic on x86)
+
+Disadvantages:
+
+  - Latency in dynamic DMA mapping, translation overhead penalty.
+  - Host software must maintain in-memory data structures for use by the IOMMU
+
+## In hardware
+
+IOMMU’s architecture is designed to accommodate a variety of system topologies.
+There can be multiple IOMMUs located at a variety of places in the system fabric.
+It can be placed at bridges inside one bus or at bridges between busses of the
+same or different types.In modern systems, it is commonly integrated with the
+PCIe root complex. It connects to the PCIe bus from downstream and to the system
+bus ( E.g. infinity fabric) from upstream. One example topology is shown below.
+
+![Example Platform Architecture](/img/iommu.png) 
+
+## In software
+
+The IOMMU is configured and controlled via two sets of registers. One in the PCI
+configuration space and another set mapped in system address space. Since the
+IOMMU appears to OS as a PCI function, it has a capability block in the PCI
+configuration space. Additionally, up to eight data structures are placed in the
+main system memory. These data structures contain information about the mapping
+of interrupts and memory for the usage of peripheral devices. Information about
+the device's memory access permissions is stored there too.
+
+## Summary
+
+IOMMU is a useful device with many advantages. Among other things, It protects a
```

Comment:
It is a very well written piece of text. What  I miss is the length and more of a practical part: Do we use IOMMU? Is it possible to extend this blogpost with some practical, low-effort for you HoWTo part??

Reply:
I'm not sure if I know any practical way that users can interact with IOMMU. As far as I understand, For 99.9% of users, it is quite an autonomous device. Unless someone writes parts of the kernel/BIOS which manage IOMMU, or writes device drivers this someone won't have to care about anything related to IOMMU. It protects PCs from e.g. DMA attacks, but users don't have to do more than turn it on to get this protection.
The only thing that I can think of is the user's ability to set up IOMMU groups. This is important e.g. while setting up VMs, but the separate post will be dedicated to IOMMU groups.
@miczyg1 Could you correct me if I'm wrong, please?

Reply:
@Sakoram I recommend the whole post: https://zirblazer.github.io/htmlfiles/coreboot.html#chapter-4.7
More to that I presented AMD IOMMU enabling topic on OSFC 2018: https://3mdeb.com/events/#Open-Source-Firmware-Conference
Finally, we should not show things from an end-user perspective only, our Independent Open Source Firmware Vendor perspective is also important. How complex and important is the enabling of those features? Are we able/capable in supporting other companies (OEM/ODM) in fixing their configuration? What about DMA attacks can we support validation?

Reply:
More:
https://looking-glass.io/
https://www.reddit.com/r/VFIO/
https://www.kernel.org/doc/Documentation/vfio.txt

Topic is extremely important.

Reply:
I added a brief description of how to enable IOMMU and what is PCIe Passthrough. If I understand correctly this should be a short introduction to IOMMU (I was told it should have around 600 words, it is ~900, now). The idea is that most of the more complex topic will be presented in future posts.

Reply:
@Sakoram I like the improvements, but please refer to some external materials, so beginners can improve their knowledge if they want.

Reply:
@pietrushnic I added a "Further reading" section. Did you read the whole post? Confirmation that everything is accurate from a technical point of view is needed.

Reply:
@Sakoram yes I did read the post and added a bunch of fixes/suggestions. The post seems to be accurate from a technical point of view. Of course, the topic is huge and in the future it would be great to explore it further.